### ISSUE TITLE
Visual Studio Win32 build fails, although platforms page says 32/64 supported

### ISSUE THREAD

**Issue by abdes** [abseil/abseil-cpp#126]
Build fails due to mishandling of int types in two library files, and dozens of the test files.

```
Build FAILED.

"...\_build\ALL_BUILD.vcxproj" (default target) (1) ->
"...\_build\third_party\abseil-cpp\absl\debugging\absl_failure_signal_handler.vcxproj" (default tar
get) (17) ->
"...\blocxxi\_build\third_party\abseil-cpp\absl\synchronization\absl_synchronization.vcxproj" (default targ
et) (21) ->
(ClCompile target) ->
  ...\third_party\abseil-cpp\absl\synchronization\internal\graphcycles.cc(281): warning C4309: 'sta
tic_cast': truncation of constant value [E:\dev\projects\blocxxi\_build\third_party\abseil-cpp\absl\synchronization\abs
l_synchronization.vcxproj]
  ...\third_party\abseil-cpp\absl\synchronization\mutex.cc(277): warning C4309: 'static_cast': trun
cation of constant value [E:\dev\projects\blocxxi\_build\third_party\abseil-cpp\absl\synchronization\absl_synchronizati
on.vcxproj]
```

Many test files fail to build as well...

**Comment by rongjiecomputer** [comment]
Works for me. Building with Visual Studio 2017 15.7 targeting x64, either with CMake + Ninja or Bazel on Windows. I don't have 32-bit compiler to test though.

Tests don't work with CMake yet. Abseil currently only supports running tests with Bazel (and tests do pass there).

**Comment by abdes** [comment]
As I said, 64bit works, 32bit build fails exactly at mutex.cc and it's a real fail. Can be easily fixed though by checking for 32bit vs 64 bit enviornment. Visual Studio can compile code both in 64 and 32. You just need to use the proper generator with cmake -G:

```
Generators
  Visual Studio 15 2017 [arch] = Generates Visual Studio 2017 project files.
                                 Optional [arch] can be "Win64" or "ARM".
  Visual Studio 14 2015 [arch] = Generates Visual Studio 2015 project files.
                                 Optional [arch] can be "Win64" or "ARM".
  Visual Studio 12 2013 [arch] = Generates Visual Studio 2013 project files.
                                 Optional [arch] can be "Win64" or "ARM".
```

The failure in 32bit compilation comes from the following cast in mutex.cc: **static_cast<uintptr_t>(0xF03A5F7BF03A5F7BLL);**

```

// We need to hide Mutexes (or other deadlock detection's pointers)
// from the leak detector.
static const uintptr_t kHideMask = static_cast<uintptr_t>(0xF03A5F7BF03A5F7BLL);
static uintptr_t MaskMu(const void *mu) {
  return reinterpret_cast<uintptr_t>(mu) ^ kHideMask;
}
```

That cast is ugly, although it has no real impact at runtime. Still, it makes the build fail:

```
....\abseil-cpp\absl\synchronization\mutex.cc(277): error C2220: warning treated as error - no 'object' file generated
....\abseil-cpp\absl\synchronization\mutex.cc(277): warning C4309: 'static_cast': truncation of constant value
```

**Comment by rongjiecomputer** [comment]
@abdes Sorry, I didn't realize your "Win32" literally mean Windows 32-bit. I normally see "Win32" as "traditional Windows API" (think "Win32" vs "UWP").

**Comment by TUNGVU1988** [comment]
yes

**Comment by derekmauro** [comment]
The compile error above should be fixed by 014f02a3eca93ef88163c3b408c86998ecf6572c, but I only tested the fixed section of code on godbolt.org with the "x86 MSVC 19 2017 RTW" compiler. I'm not sure what flags you need to pass to cl.exe to force 32-bit mode. There may be other errors.

I'm going to close this bug. Re-open if there is another 32-bit error. Patches are also welcome.

**Comment by abdes** [comment]
I reviewed your changes and they look good. Thanks for the quick turnaround.
